### DISCUSSION TITLE
atomWithObservable hangs Suspense with observable chains

### DISCUSSION THREAD

**Discussion by dakom** [pmndrs/jotai#2007] [Bug report] [answered]
See the comments [in this test](https://github.com/pmndrs/jotai/pull/2008):

```
it('writable count state without initial value and rxjs chain', async () => {
  const single$ = new Subject<number>()
  const double$ = single$.pipe(map((n) => n * 2))
  const singleAtom = atomWithObservable(() => single$)

  // this currently breaks:
  //const doubleAtom = atomWithObservable(() => double$)

  // this works, however:
  const doubleAtom = atom(async (get) => {
    const n = await get(singleAtom)
    return n * 2
  })

  const CounterValue = () => {
    const single = useAtomValue(singleAtom)
    const double = useAtomValue(doubleAtom)
    return (
      <>
        single: {single}, double: {double}
      </>
    )
  }

  const CounterButton = () => {
    return <button onClick={() => single$.next(2)}>button</button>
  }

  const { findByText, getByText } = render(
    <StrictMode>
      <Suspense fallback="loading">
        <CounterValue />
      </Suspense>
      <CounterButton />
    </StrictMode>
  )

  await findByText('loading')

  fireEvent.click(getByText('button'))
  await findByText('single: 2, double: 4')

  act(() => single$.next(3))
  await findByText('single: 3, double: 6')
})
```

**Comment by dai-shi** [comment]
I wonder if something like this works:
```js
const singleAtom = atomWithObservable(() => single$)
const doubleAtom = atomWithObservable(() => double$)
const singleAndDoubleAtom = atom((get) => Promise.all([get(singleAtom), get(doubleAtom)]))

  const [single, double] = useAtomValue(singleAndDoubleAtom)
```

**Reply by dakom** [reply]
I've updated [the PR](https://github.com/pmndrs/jotai/pull/2008) so that it has the suggested workaround in comments, and so that it now fails out of the box when running the test

**Reply by dai-shi** [reply]
If I understand it correctly, it's not a bug in `atomWithObservable`. It can be reproduced with any async atoms.
Would you be interested in updating your PR to try reproducing the bug without `atomWithObservable`?
I'm not sure if it's a bug. It's rather a limitation than a bug. We could probably reproduce the behavior without Jotai.

FYI, this is kind of a known problem (waterfall), and we develop a library to mitigate it: https://github.com/jotaijs/jotai-suspense `usePrepareAtoms`

**Reply by dai-shi** [reply]
> It can be reproduced with any async atoms.

https://github.com/pmndrs/jotai/discussions/2007#discussioncomment-6330199 Now, I'm not very confident. More investigation is appreciated.

**Comment by dai-shi** [comment]
@dakom

> though interestingly, this breaks it again:
> 
> ```
>   const singleAndDoubleAtom = atom(async (get) => {
>     const single = await get(singleAtom)
>     const double = await get(doubleAtom)
>     return [single, double]
>   })
> ```

Oh, that's interesting. It should be solved, if possible.
If we await doubleAtom first, does the issue go away? (Then, it might be an issue specific to `atomWithObservable`.)

**Comment by dakom** [comment]
awaiting `doubleAtom` first does _not_ fix the issue

`usePrepareAtoms` _does_ fix the issue, I opened a PR on `jotai-suspense` to have this test case: https://github.com/jotaijs/jotai-suspense/pull/3 (if you leave out the `Prepare` component, it fails)

**Reply by dai-shi** [reply]
Thanks for the confirming.

I haven't yet had a clear understanding of where the problem is, but my rough rules are:
- If it's atomWithObservable only issue (or use case), it sounds tough (= not easy to fix, and rather a limitation meanwhile, but anyone can tackle)
- If it's an issue with bare async atoms, we should basically fix it:
  - if it's an issue with the store api, it's our control and we should be able to fix it.
  - if it's the issue with `useAtom`, it might be because of React behavior, and we might not be able to fix (Like the usePrepareAtoms case.)

**Comment by tmkx** [comment, marked as answer]
for now, it seems work as expected...
```tsx
const CounterValue = () => {
  const single = useAtomValue(singleAtom) // throw promise until subject$.next
  const double = useAtomValue(doubleAtom) // `double` will mount **after** subject$.next, so it will make Suspense still in the fallback state
  return (
    <>
      single: {single}, double: {double}
    </>
  )
}
```

**Reply by dai-shi** [reply]
Hmm, #2008 still fails. Not sure if something is different. Can you take a look?

(edit: I misunderstood the original comment. 😓 )

**Reply by tmkx** [reply]
there are some ways to avoid:

1. recap: combine two obserableAtom into one atom, which ensures that both of them will be mounted
```ts
const singleAndDoubleAtom = atom((get) => Promise.all([get(singleAtom), get(doubleAtom)]))
```

2. assign `initialValue`
```ts
const singleAtom = atomWithObservable(() => single$, { initialValue: 0 })
const doubleAtom = atomWithObservable(() => double$, { initialValue: 0 })
```

3. replace `Subject` with `BehaviorSubject`
```ts
const single$ = new BehaviorSubject<number>(0)
```

---

```ts
const singleAndDoubleAtom = atom(async (get) => {
  const single = await get(singleAtom)
  const double = await get(doubleAtom)
  return [single, double]
})
```

this will not work because it's waiting for `subject$.next` sequentially, it needs two `subject$.next` calls to fullfill it.

**Reply by dai-shi** [reply]
Okay, so we think it's failing by design. Yeah, that makes sense. Thanks for the insight!

**Reply by tmkx** [reply]
The vanilla one is passed: https://github.com/pmndrs/jotai/pull/2008/files#diff-e0e645d4fdf46246c5e371cc26ad6cc82b7191a3b0de2a15da2f1ea926408418R847-R874

another one is work as expected so far(?)

**Reply by dai-shi** [reply]
Yeah, should we merge the working one?

**Reply by tmkx** [reply]
LGTM
